### PR TITLE
Marian's Mongo-Angular-Bank

### PR DESCRIPTION
**Pull request by NewestUser**

### CODE REVIEW

File: bank/src/main/java/com/clouway/bricky/adapter/http/HttpServletModule.java [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+import com.google.inject.servlet.ServletModule;
+
+/**
+ * @author Marian Zlatev <mzlatev91@gmail.com>
+ */
+public class HttpServletModule extends ServletModule {
+
+  @Override
+  protected void configureServlets() {
+    filter("*").through(SecurityFilter.class);
```
Comment (freekman) [inline bank/src/main/java/com/clouway/bricky/adapter/http/HttpServletModule.java]: Does your filter have to listen on all files.

File: bank/src/tests/com/clouway/bricky/core/sesion/UserSessionManagerTest.java [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+
+    assertFalse(manager.isUserSessionExpired());
+  }
+
+  @Test
+  public void lookupNonExistingSession() throws Exception {
+    context.checking(new Expectations() {{
+      oneOf(session).getSid();
+      will(returnValue(Optional.absent()));
+//      oneOf(request).getCookies();
```
Comment (freekman) [inline bank/src/tests/com/clouway/bricky/core/sesion/UserSessionManagerTest.java]: remove commented code

File: bank/src/tests/com/clouway/bricky/core/sesion/UserSessionManagerTest.java [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+  }
+
+  @Test
+  public void refreshCurrentUserSession() throws Exception {
+    context.checking(new Expectations() {{
+      oneOf(session).getSid();
+      will(returnValue(Optional.of("123")));
+//      oneOf(request).getCookies();
+//      will(returnValue(dummySession()));
+      oneOf(repository).refreshSession("123");
```
Comment (freekman) [inline bank/src/tests/com/clouway/bricky/core/sesion/UserSessionManagerTest.java]: .

File: bank/src/tests/com/clouway/bricky/core/sesion/UserSessionManagerTest.java [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+  @Test
+  public void removeRedundantSession() throws Exception {
+    context.checking(new Expectations() {{
+      oneOf(session).getSid();
+      will(returnValue(dummySession()));
+      oneOf(session).detach();
+//      oneOf(request).getCookies();
+//      will(returnValue(dummySession()));
+//      oneOf(response).addCookie(with(any(Cookie.class)));
+      oneOf(repository).clearSession(with(any(String.class)));
```
Comment (freekman) [inline bank/src/tests/com/clouway/bricky/core/sesion/UserSessionManagerTest.java]: .

File: bank/src/tests/com/clouway/bricky/core/sesion/UserSessionManagerTest.java [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+    }});
+
+    manager.closeUserSession();
+  }
+
+  @NotNull
+  private User dummyUser() {
+    return new User("Marian", "pswd");
+  }
+
```
Comment (freekman) [inline bank/src/tests/com/clouway/bricky/core/sesion/UserSessionManagerTest.java]: Why are you using @NotNull new User will never be null

File: bank/src/main/java/com/clouway/bricky/core/db/balance/BalanceRepository.java [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+ * @author Marian Zlatev <mzlatev91@gmail.com>
+ */
+public interface BalanceRepository {
+
+  CurrentUser depositToCurrentUser(double amount);
+
+  CurrentUser withdrawFromCurrentUser(double amount) throws FundDeficitException;
+
+  CurrentUser getCurrentUser();
+}
```
Comment (freekman) [inline bank/src/main/java/com/clouway/bricky/core/db/balance/BalanceRepository.java]: can you deposit or withdraw from some one else then current user

File: bank/src/main/java/com/clouway/bricky/core/sesion/SessionManager.java [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+ */
+public interface SessionManager {
+
+  /**
+   * Attach session id for specified user.
+   *
+   * @param user that will have a session opened for.
+   */
+  void openSessionFor(User user);
+
```
Comment (freekman) [inline bank/src/main/java/com/clouway/bricky/core/sesion/SessionManager.java:16]: I think that openSession is better

File: bank/src/main/java/com/clouway/bricky/adapter/http/HttpServletModule.java (hunk 1 above)
Comment (NewestUser) [inline bank/src/main/java/com/clouway/bricky/adapter/http/HttpServletModule.java]: This way I guarantee that all my rest service's are protected. When I create a new one I won't have to come and add it to the filter. The only accessible pages are Login and Register for now. If my logic changes maybe I would have to exclude some pages not to be filtered.

File: bank/src/main/java/com/clouway/bricky/core/db/balance/BalanceRepository.java (hunk 6 above)
Comment (NewestUser) [inline bank/src/main/java/com/clouway/bricky/core/db/balance/BalanceRepository.java]: No I can't perform operations on user other than CurrentUser. Currently my back is only working with the loged in user. If I have to do transactions from one account to another maybe i will have a TransactionRepository or something like that that will perform transactions from one used account to another.

File: bank/src/main/java/com/clouway/bricky/core/sesion/SessionManager.java (hunk 7 above)
Comment (NewestUser) [inline bank/src/main/java/com/clouway/bricky/core/sesion/SessionManager.java:16]: Renamed to openSession

File: bank/src/tests/com/clouway/bricky/core/sesion/UserSessionManagerTest.java (hunk 2 above)
Comment (NewestUser) [inline bank/src/tests/com/clouway/bricky/core/sesion/UserSessionManagerTest.java]: Comments removed.

File: bank/src/tests/com/clouway/bricky/core/sesion/UserSessionManagerTest.java (hunk 5 above)
Comment (NewestUser) [inline bank/src/tests/com/clouway/bricky/core/sesion/UserSessionManagerTest.java]: IDEA added @NotNull when I extracted declaration of User to a method.
@NotNull is used to label a method that will never return null.
Maybe in the tests this isn't necessary and I will remove it.

File: bank/src/main/java/com/clouway/bricky/adapter/http/HttpServletModule.java (hunk 1 above)
Comment (freekman) [inline bank/src/main/java/com/clouway/bricky/adapter/http/HttpServletModule.java]: ok

**Comment by freekman** [comment]
@NewestUser  LGTM

File: bank/src/main/java/com/clouway/bricky/GuiceListener.java [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+
+/**
+ * @author Marian Zlatev (mzlatev91@gmail.com)
+ */
+public class GuiceListener extends GuiceServletContextListener {
+  @Override
+  protected Injector getInjector() {
+    return Guice.createInjector(new HttpServletModule(), new HttpModule(), new BrickModule(), new PersistentDbModule());
+  }
+}
```
Comment (EmilGeorgiev) [inline bank/src/main/java/com/clouway/bricky/GuiceListener.java]: This class is unnecessary. You should user `Guice.createInjector(new .....)` in main method of theapplication.

File: bank/src/main/java/com/clouway/bricky/adapter/http/AmountDTO.java [hunk 9]
```diff
@@ -0,0 +1,13 @@
+package com.clouway.bricky.adapter.http;
+
+/**
+ * @author Marian Zlatev <mzlatev91@gmail.com>
+ */
+class AmountDTO {
+
+  public double amount;
```
Comment (EmilGeorgiev) [inline bank/src/main/java/com/clouway/bricky/adapter/http/AmountDTO.java]: this should be final. Also we skip `public` because class is package private.

Use `Double`class instead `double` variable.

We use `javax.validation` library 

``` javascript
<dependency>
      <groupId>javax.validation</groupId>
      <artifactId>validation-api</artifactId>
      <version>1.0.0.GA</version>
</dependency>
```

for validate DTO field. For Example:

``` javascript
class AmountDTO {
  @NotNull(message="Amount is required")
  final Double amount;

  AmountDTO(Dounble amount) {
   this.amount = amount;
  }
}
```

File: bank/src/main/java/com/clouway/bricky/adapter/http/BalanceService.java [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+  }
+
+  @Post
+  @At("/deposit")
+  public Reply<?> deposit(Request request) {
+    AmountDTO dto = request.read(AmountDTO.class).as(Json.class);
+
+    if (dto.amount < 0) {
+      return Reply.with("Operation failed.").status(SC_BAD_REQUEST).as(Json.class);
+    }
```
Comment (EmilGeorgiev) [inline bank/src/main/java/com/clouway/bricky/adapter/http/BalanceService.java]: In REST service should not have logic. If use `javax.validation.constraints.Pattern` annotation we can validate amount to match to same regex (in this case pattern for numbers greater then zero). And this logic will be unnecessary.

File: bank/src/main/java/com/clouway/bricky/adapter/http/BalanceService.java [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+  }
+
+  @Post
+  @At("/withdraw")
+  public Reply<?> withdraw(Request request) {
+    AmountDTO dto = request.read(AmountDTO.class).as(Json.class);
+
+    if (dto.amount < 0) {
+      return Reply.with("Operation failed.").status(SC_FORBIDDEN).as(Json.class);
+    }
```
Comment (EmilGeorgiev) [inline bank/src/main/java/com/clouway/bricky/adapter/http/BalanceService.java]: .

File: bank/src/main/java/com/clouway/bricky/adapter/http/BalanceService.java [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+    try {
+      CurrentUser user = repository.withdrawFromCurrentUser(dto.amount);
+      return Reply.with(user.balance).status(SC_CREATED).as(Json.class);
+    } catch (FundDeficitException e) {
+      return Reply.with("Operation failed.").status(SC_FORBIDDEN).as(Json.class);
+    }
```
Comment (EmilGeorgiev) [inline bank/src/main/java/com/clouway/bricky/adapter/http/BalanceService.java]: REST service should not catch exceptions. This is not his responsibility.but for now leave him.

File: bank/src/main/java/com/clouway/bricky/adapter/http/AmountDTO.java (hunk 9 above)
Comment (NewestUser) [inline bank/src/main/java/com/clouway/bricky/adapter/http/AmountDTO.java]: I agree about the public modifier.
I will change it to Double and will read about javax.validation.

Currently I can not make the field final because the transporter class that we use does not pass parameters through constructor, instead a setter method is used and thus final is not permitted.

File: bank/src/main/java/com/clouway/bricky/GuiceListener.java (hunk 8 above)
Comment (NewestUser) [inline bank/src/main/java/com/clouway/bricky/GuiceListener.java]: Removed class and webinf folder. Guice injector is created when instantiating jetty server.

File: bank/src/main/java/com/clouway/bricky/adapter/http/BalanceService.java (hunk 12 above)
Comment (NewestUser) [inline bank/src/main/java/com/clouway/bricky/adapter/http/BalanceService.java]: Note taken.

File: bank/src/main/java/com/clouway/bricky/MainServer.java [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+    servletContextHandler.addFilter(GuiceFilter.class, "/*", EnumSet.allOf(DispatcherType.class));
+
+    // You MUST add DefaultServlet or your server will always return 404s
+    servletContextHandler.addServlet(DefaultServlet.class, "/");
+    servletContextHandler.setResourceBase("frontend");
+    return server;
+  }
+
+  private static PropertyReader getPropertyReader(String[] args) {
+    String propName;
```
Comment (EmilGeorgiev) [inline bank/src/main/java/com/clouway/bricky/MainServer.java]: rename propName to configurationFile

File: bank/src/main/java/com/clouway/bricky/MainServer.java [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+    if (args != null && args.length != 0) {
+      propName = args[0];
+      System.out.println("PROP FILE: " + propName);
+    } else {
+      propName = "../config.properties";
+    }
```
Comment (EmilGeorgiev) [inline bank/src/main/java/com/clouway/bricky/MainServer.java:55]:
``` javascript
String configurationFile = configuration.properties;

if(args.lenght != 0) {
  configurationFile = args[0];
}
```

Thus you avoid else statement.

File: bank/src/main/java/com/clouway/bricky/MainServer.java [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+    return server;
+  }
+
+  private static PropertyReader getPropertyReader(String[] args) {
+    String propName;
+    if (args != null && args.length != 0) {
+      propName = args[0];
+      System.out.println("PROP FILE: " + propName);
+    } else {
+      propName = "../config.properties";
```
Comment (EmilGeorgiev) [inline bank/src/main/java/com/clouway/bricky/MainServer.java]: configuration.properties files should be in directory of project. i think that `../` is unnecessary

File: bank/src/main/java/com/clouway/bricky/PropertyReader.java [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Properties;
+
+/**
+ * @author Marian Zlatev <mzlatev91@gmail.com>
+ */
+public class PropertyReader {
+
+  private Properties properties;
```
Comment (EmilGeorgiev) [inline bank/src/main/java/com/clouway/bricky/PropertyReader.java]: make final

File: bank/src/main/java/com/clouway/bricky/core/PersistentDbModule.java [hunk 17]
```diff
@@ -0,0 +1,46 @@
+package com.clouway.bricky.core;
```
Comment (EmilGeorgiev) [inline bank/src/main/java/com/clouway/bricky/core/PersistentDbModule.java]: Persistent module should be in `persistent` package not in `core`.

File: bank/src/main/java/com/clouway/bricky/core/PersistentDbModule.java [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+public class PersistentDbModule extends AbstractModule {
+  private PropertyReader properties;
+
+  public PersistentDbModule(PropertyReader reader) {
+    this.properties = reader;
+  }
+
+  @Override
+  protected void configure() {
+    bind(UserRepository.class).to(MongoUserRepository.class);
```
Comment (EmilGeorgiev) [inline bank/src/main/java/com/clouway/bricky/core/PersistentDbModule.java]: `bind(UserRepository.class).to(MongoUserRepository.class).in(Singleton.class);`

File: bank/src/main/java/com/clouway/bricky/core/Registry.java [hunk 19]
```diff
@@ -0,0 +1,12 @@
+package com.clouway.bricky.core;
+
+import com.clouway.bricky.core.user.User;
+
+/**
+ * @author Marian Zlatev <mzlatev91@gmail.com>
+ */
+public interface Registry {
```
Comment (EmilGeorgiev) [inline bank/src/main/java/com/clouway/bricky/core/Registry.java:10]: Add documentation

File: bank/src/main/java/com/clouway/bricky/core/Registry.java [hunk 20]
```diff
@@ -0,0 +1,12 @@
...
+
+  void authorize(User user) throws UnauthorizedException;
```
Comment (EmilGeorgiev) [inline bank/src/main/java/com/clouway/bricky/core/Registry.java]: .

File: bank/src/main/java/com/clouway/bricky/MainServer.java (hunk 13 above)
Comment (NewestUser) [inline bank/src/main/java/com/clouway/bricky/MainServer.java]: variable renamed

File: bank/src/main/java/com/clouway/bricky/MainServer.java (hunk 15 above)
Comment (NewestUser) [inline bank/src/main/java/com/clouway/bricky/MainServer.java]: done

File: bank/src/main/java/com/clouway/bricky/MainServer.java (hunk 14 above)
Comment (NewestUser) [inline bank/src/main/java/com/clouway/bricky/MainServer.java:55]: String is initialized and else statement removed.

File: bank/src/main/java/com/clouway/bricky/PropertyReader.java (hunk 16 above)
Comment (NewestUser) [inline bank/src/main/java/com/clouway/bricky/PropertyReader.java]: done

File: bank/src/main/java/com/clouway/bricky/core/PersistentDbModule.java (hunk 17 above)
Comment (NewestUser) [inline bank/src/main/java/com/clouway/bricky/core/PersistentDbModule.java]: package moved

File: bank/src/main/java/com/clouway/bricky/core/PersistentDbModule.java (hunk 18 above)
Comment (NewestUser) [inline bank/src/main/java/com/clouway/bricky/core/PersistentDbModule.java]: Repositories are bind as Singleton.

File: bank/src/main/java/com/clouway/bricky/core/Registry.java (hunk 19 above)
Comment (NewestUser) [inline bank/src/main/java/com/clouway/bricky/core/Registry.java:10]: done

File: bank/src/main/java/com/clouway/bricky/core/Registry.java (hunk 20 above)
Comment (NewestUser) [inline bank/src/main/java/com/clouway/bricky/core/Registry.java]: done
